mountpoint-s3-crt/src/s3/pool.rs

```diff
@@ -0,0 +1,432 @@
+    unsafe { crt_pool.leak() }
+}
+
+fn drop_pool_factory<PoolFactory: MemoryPoolFactory>(factory_ptr: *mut libc::c_void) {
+    // SAFETY: `factory_ptr` was leaked in `CrtBufferPoolFactory::new`.
+    _ = unsafe { Pin::new_unchecked(Box::from_raw(factory_ptr as *mut PoolFactory)) };
+}
+
+/// Internal wrapper to bridge the [MemoryPool] implementation to
+/// the `aws_s3_buffer_pool` to provide to the CRT.
+///
+/// Instances of this type also hold the vtables to set up both
+/// the `aws_s3_buffer_pool` itself and the `aws_s3_buffer_ticket`s
+/// it returns. Notably, all the functions in the vtables are generic
+/// in the same [MemoryPool] implementation as [CrtBufferPool], so
+/// that the CRT can handle different implementations and there is
+/// no need for dynamic dispatch on the Rust side.
+struct CrtBufferPool<Pool: MemoryPool> {
+    /// Inner struct to pass to CRT functions.
+    inner: aws_s3_buffer_pool,
+    /// [MemoryPool] implementation.
+    pool: Pool,
+    /// Holds the vtable to point to in `inner`.
+    pool_vtable: aws_s3_buffer_pool_vtable,
+    /// Holds the vtable for the `aws_s3_buffer_ticket` instances.
+    ticket_vtable: aws_s3_buffer_ticket_vtable,
+    /// CRT allocator.
+    allocator: Allocator,
+    /// Pin this struct because inner.impl_ will be a pointer to this object.
+    _pinned: PhantomPinned,
+}
+
+impl<Pool: MemoryPool> CrtBufferPool<Pool> {
+    fn new(pool: Pool, allocator: Allocator) -> Pin<Box<Self>> {
+        // `inner` will be initialized after pinning because its fields require pinned addresses.
+        let mut crt_pool = Box::pin(CrtBufferPool {
+            inner: Default::default(),
+            pool,
+            pool_vtable: aws_s3_buffer_pool_vtable {
+                reserve: Some(pool_reserve::<Pool>),
+                trim: Some(pool_trim::<Pool>),
+                acquire: None,
+                release: None,
+            },
+            ticket_vtable: aws_s3_buffer_ticket_vtable {
+                claim: Some(ticket_claim::<Pool::Buffer>),
+                acquire: None,
+                release: None,
+            },
+            allocator,
+            _pinned: Default::default(),
+        });
+
+        // Set up the vtable and `impl_` to the pinned addresses (self-referential) and initialize ref-counting.
+        // SAFETY: We're setting up the struct to be self-referential, and we're not moving out
+        // of the struct, so the unchecked deref of the pinned pointer is okay.
+        unsafe {
+            let pool_ref = Pin::get_unchecked_mut(Pin::as_mut(&mut crt_pool));
+            pool_ref.inner.vtable = &raw mut pool_ref.pool_vtable;
+            pool_ref.inner.impl_ = pool_ref as *mut CrtBufferPool<Pool> as *mut libc::c_void;
+            aws_ref_count_init(
+                &mut pool_ref.inner.ref_count,
+                &mut pool_ref.inner as *mut aws_s3_buffer_pool as *mut libc::c_void,
+                Some(pool_destroy::<Pool>),
```

Comment:
does it mean that "outer" struct `CrtBufferPool` relies on refcounter of it's field `inner` to get destroyed eventually?